stimpacks/deploy/config.go

```diff
@@ -0,0 +1,352 @@
+				&EnvironmentVar{Name: "VAULT_TOKEN", Value: vaultToken},
+				&EnvironmentVar{Name: "DEPLOY_ENVIRONMENT", Value: environment.Name},
+				&EnvironmentVar{Name: "DEPLOY_INSTANCE", Value: instance.Name},
+				&EnvironmentVar{Name: "DEPLOY_CLUSTER", Value: instance.Spec.Kubernetes.Cluster},
+			}...)
+
+			// Generate the Kube config secret
+			var stimSecrets []*v2e.SecretItem
+			if instance.Spec.Kubernetes.ServiceAccount != "" {
+				secretMap := make(map[string]string)
+				secretMap["CLUSTER_SERVER"] = "cluster-server"
+				secretMap["CLUSTER_CA"] = "cluster-ca"
+				secretMap["USER_TOKEN"] = "user-token"
+				stimSecrets = append(stimSecrets, &v2e.SecretItem{
+					SecretPath: fmt.Sprintf("secret/kubernetes/%s/%s/kube-config", instance.Spec.Kubernetes.Cluster, instance.Spec.Kubernetes.ServiceAccount),
+					SecretMaps: secretMap,
+				})
+			} else {
+				d.log.Fatal("Kubernetes service account required but not provided")
+			}
+
+			// Add stim envs/secrets and ensure no reserved env vars have been set
+			d.finalizeEnv(instance, stimEnvs, stimSecrets)
+		}
+	}
+
+	// Determine the full directory path
+	d.config.Deployment.fullDirectoryPath = filepath.Join(filepath.Dir(d.config.configFilePath), d.config.Deployment.Directory)
+
+}
+
+// Generate the list of reserved env var names
+func (d *Deploy) finalizeEnv(instance *Instance, stimEnvs []*EnvironmentVar, stimSecrets []*v2e.SecretItem) {
+
+	// Generate the list of reserved env var names (additionally SECRET_CONFIG as we'll add that one at the end)
+	reservedVarNames := []string{"SECRET_CONFIG"}
+	for _, s := range stimEnvs {
+		reservedVarNames = append(reservedVarNames, s.Name)
+	}
+	for _, s := range stimSecrets {
+		for m := range s.SecretMaps {
+			reservedVarNames = append(reservedVarNames, m)
+		}
+	}
+
+	// Exit if any user-provided environment vars conflict with reserved ones
+	for _, e := range instance.Spec.EnvironmentVars {
+		if utils.Contains(reservedVarNames, e.Name) {
+			d.log.Fatal("Reserved environment variable name '{}' found in config", e.Name)
+		}
+	}
+	for _, s := range instance.Spec.Secrets {
+		for m := range s.SecretMaps {
+			if utils.Contains(reservedVarNames, m) {
+				d.log.Fatal("Reserved environment variable name '{}' found in config", m)
+			}
+		}
+	}
+
+	// Combine our secrets
+	instance.Spec.Secrets = append(instance.Spec.Secrets, stimSecrets...)
+
+	// Create the secret config
+	secretConfig, err := d.makeSecretConfig(instance)
+	if err != nil {
+		panic(err)
+	}
+	stimEnvs = append(stimEnvs, &EnvironmentVar{Name: "SECRET_CONFIG", Value: secretConfig})
+
+	// Combine our env vars
+	instance.Spec.EnvironmentVars = append(instance.Spec.EnvironmentVars, stimEnvs...)
+
+}
+
+// mergeEnvVars is used to merge environment variable configuration at the various levels it can be set at
+func mergeEnvVars(instance []*EnvironmentVar, environment []*EnvironmentVar, global []*EnvironmentVar) []*EnvironmentVar {
```

Comment:
Not sure this is worth changing but I think you want to use something more like a Set for EnvironmentVar.  It makes finding collisions, merging and diffing much easier.

"github.com/deckarep/golang-set"

Reply:
Yeah - I don't love the way I'm doing the merging of envs and secrets so I wouldn't be opposed to changing it.  I'm not familiar with Python's "set" so I'll have to read up on it but I agree that this can be done in a better way.